File: frontend/src/utils/logger/sensitive.ts
const SENSITIVE_KEYS = new Set([
  'password', 'passwd', 'pwd', 'secret', 'token', 'access_token',
  'refresh_token', 'api_key', 'apikey', 'private_key', 'credit_card',
  'card_number', 'cvv', 'ssn', 'id_card', 'bank_card', 'authorization',
  'cookie',
])

const MASK = '******'

const PHONE_RE = /1[3-9]\d{9}/g
const ID_CARD_RE = /[1-9]\d{5}(?:19|20)\d{2}(?:0[1-9]|1[0-2])(?:0[1-9]|[12]\d|3[01])\d{3}[\dXx]/g
const BANK_CARD_RE = /\b\d{16,19}\b/g
const EMAIL_RE = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g

function maskStringPatterns(value: string): string {
  if (typeof value !== 'string') return value
  value = value.replace(PHONE_RE, m => m.slice(0, 3) + MASK + m.slice(-2))
  value = value.replace(ID_CARD_RE, m => m.slice(0, 3) + MASK + m.slice(-1))
  value = value.replace(BANK_CARD_RE, m => m.slice(0, 4) + MASK + m.slice(-4))
  value = value.replace(EMAIL_RE, m => {
    const parts = m.split('@')
    return parts[0].slice(0, 2) + MASK + '@' + parts[1]
  })
  return value
}

export function sanitizeData(data: unknown, depth = 0): unknown {
  if (depth > 10) return '...'
  if (data === null || data === undefined) return data
  if (typeof data === 'boolean') return data
  if (typeof data === 'number') return data
  if (typeof data === 'string') return maskStringPatterns(data)
  if (Array.isArray(data)) return data.map(item => sanitizeData(item, depth + 1))
  if (typeof data === 'object') {
    const result: Record<string, unknown> = {}
    for (const [key, value] of Object.entries(data as Record<string, unknown>)) {
      const keyLower = key.toLowerCase()
      if (SENSITIVE_KEYS.has(keyLower)) {
        result[key] = MASK
      } else if (typeof value === 'string') {
        result[key] = maskStringPatterns(value)
      } else {
        result[key] = sanitizeData(value, depth + 1)
      }
    }
    return result
  }
  return String(data)
}

File: backend/logging/sensitive.py
import re
from typing import Any


_SENSITIVE_KEYS = {
    'password', 'passwd', 'pwd', 'secret', 'token', 'access_token',
    'refresh_token', 'api_key', 'apikey', 'private_key', 'credit_card',
    'card_number', 'cvv', 'ssn', 'id_card', 'bank_card', 'phone',
    'mobile', 'email', 'authorization', 'cookie',
}

_MASK = '******'

_PHONE_PATTERN = re.compile(r'1[3-9]\d{9}')
_ID_CARD_PATTERN = re.compile(r'[1-9]\d{5}(?:19|20)\d{2}(?:0[1-9]|1[0-2])(?:0[1-9]|[12]\d|3[01])\d{3}[\dXx]')
_BANK_CARD_PATTERN = re.compile(r'\b\d{16,19}\b')
_EMAIL_PATTERN = re.compile(r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}')


def mask_value(value: str) -> str:
    if not isinstance(value, str) or len(value) < 3:
        return _MASK
    return value[:2] + _MASK + value[-1:]


def mask_phone(value: str) -> str:
    if _PHONE_PATTERN.fullmatch(value):
        return value[:3] + _MASK + value[-2:]
    return value


def mask_id_card(value: str) -> str:
    if _ID_CARD_PATTERN.fullmatch(value):
        return value[:3] + _MASK + value[-1:]
    return value


def mask_bank_card(value: str) -> str:
    if _BANK_CARD_PATTERN.fullmatch(value):
        return value[:4] + _MASK + value[-4:]
    return value


def mask_email(value: str) -> str:
    if _EMAIL_PATTERN.fullmatch(value):
        parts = value.split('@')
        if len(parts) == 2:
            return parts[0][:2] + _MASK + '@' + parts[1]
    return value


def mask_string_patterns(value: str) -> str:
    if not isinstance(value, str):
        return value
    value = _PHONE_PATTERN.sub(lambda m: m.group()[:3] + _MASK + m.group()[-2:], value)
    value = _ID_CARD_PATTERN.sub(lambda m: m.group()[:3] + _MASK + m.group()[-1:], value)
    value = _BANK_CARD_PATTERN.sub(lambda m: m.group()[:4] + _MASK + m.group()[-4:], value)
    value = _EMAIL_PATTERN.sub(lambda m: m.group()[:2] + _MASK + '@' + m.group().split('@')[1], value)
    return value


def sanitize_data(data: Any, depth: int = 0) -> Any:
    if depth > 10:
        return '...'
    if data is None:
        return None
    if isinstance(data, bool):
        return data
    if isinstance(data, (int, float)):
        return data
    if isinstance(data, str):
        return mask_string_patterns(data)
    if isinstance(data, (list, tuple)):
        return [sanitize_data(item, depth + 1) for item in data]
    if isinstance(data, dict):
        result = {}
        for key, value in data.items():
            key_lower = key.lower() if isinstance(key, str) else str(key).lower()
            if key_lower in _SENSITIVE_KEYS:
                result[key] = _MASK
            elif isinstance(value, str):
                result[key] = mask_string_patterns(value)
            else:
                result[key] = sanitize_data(value, depth + 1)
        return result
    if hasattr(data, '__dict__'):
        return sanitize_data(data.__dict__, depth + 1)
    return str(data)
